router-tests/testenv/testenv.go

```diff
@@ -875,6 +883,81 @@ func (e *Environment) MakeGraphQLRequest(request GraphQLRequest) (*TestResponse,
 	}, nil
 }
 
+func (e *Environment) MakeGraphQLRequestAsMultipartForm(request GraphQLRequest) (*TestResponse, error) {
+	data, err := json.Marshal(request)
+	require.NoError(e.t, err)
+
+	formValues := map[string]io.Reader{
+		"operations": bytes.NewReader(data),
+		"map":        strings.NewReader(`{ "0": ["variables.file"] }`),
+		"0":          bytes.NewReader(request.Files[0]),
+	}
+	multipartBody, contentType, err := multipartBytes(formValues)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequestWithContext(e.Context, http.MethodPost, e.GraphQLRequestURL(), &multipartBody)
+	if err != nil {
+		return nil, err
+	}
+
+	if request.Header != nil {
+		req.Header = request.Header
+	}
+	req.Header.Add("Content-Type", contentType)
+
+	resp, err := e.RouterClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	resp.Body = io.NopCloser(bytes.NewReader(buf.Bytes()))
+	body := buf.String()
+
+	return &TestResponse{
+		Body:     strings.TrimSpace(body),
+		Response: resp,
+		Proto:    resp.Proto,
+	}, nil
+}
+
+func multipartBytes(values map[string]io.Reader) (bytes.Buffer, string, error) {
+	var err error
+	var b bytes.Buffer
+	w := multipart.NewWriter(&b)
+	for key, r := range values {
+		var fw io.Writer
+		if x, ok := r.(io.Closer); ok {
+			defer x.Close()
```

Comment:
Defer in a loop is forbidden. You can extract in a function and call defer there or call close everywhere.

Reply:
Done